src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -1205,28 +1170,48 @@ public StringBuilder Append(char[]? value)
         {
             if (value?.Length > 0)
             {
-                unsafe
-                {
-                    fixed (char* valueChars = &value[0])
-                    {
-                        Append(valueChars, value.Length);
-                    }
-                }
+                Append(value.AsSpan());
             }
             return this;
         }
 
         public StringBuilder Append(ReadOnlySpan<char> value)
         {
-            if (value.Length > 0)
+            // This is where we can check if adding value will put us over m_MaxCapacity.
+            // We are doing the check here to prevent the corruption of the StringBuilder.
+            int newLength = Length + value.Length;
+            if (newLength > m_MaxCapacity || newLength < value.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(value), SR.ArgumentOutOfRange_LengthGreaterThanCapacity);
+            }
```

Comment:
Can this method instead be structured as:
```C#
// This case is so common we want to optimize for it heavily.
if (value.TryCopyTo(m_ChunkChars.AsSpan(m_ChunkLength)))
{
    m_ChunkLength += value.Length;
}
else if (value.Length != 0)
{
    ... // the above argument/state validation
    ... // the contents of the else block below
}
return this;
```
?

Reply:
Done (don't know why GitHub doesn't show "Outdated").